MdeModulePkg/Library/MemoryBucketLib/MemoryBucketLib.c

```diff
@@ -0,0 +1,530 @@
+                          in the PEI memory bucket structure.
+  @retval    FALSE        The type of memory we are allocating is not
+                          included in the PEI memory bucket structure.
+**/
+BOOLEAN
+EFIAPI
+IsRuntimeType (
+  IN       EFI_MEMORY_TYPE  MemoryType
+  )
+{
+  UINTN  Index;
+
+  // If Runtime Buckets are disabled then deny memory bucket operations.
+  if (mMemoryBucketsDisabled) {
+    return FALSE;
+  }
+
+  for (Index = 0; Index < PEI_BUCKETS; Index++) {
+    if (MemoryType == mMemoryTypes[Index]) {
+      return TRUE;
+    }
+  }
+
+  return FALSE;
+}
+
+/**
+  Function that checks if PEI memory buckets are enabled.
+
+  @retval    TRUE         PEI memory buckets are enabled
+  @retval    FALSE        PEI memory buckets are disabled
+**/
+BOOLEAN
+EFIAPI
+AreMemoryBucketsEnabled (
+  VOID
+  )
+{
+  if (mMemoryBucketsDisabled) {
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+/**
+  Internal function to build a HOB for the memory allocation.
+  It will search and reuse the unused(freed) memory allocation HOB,
+  or build memory allocation HOB normally if no unused(freed) memory allocation HOB found.
+  @param[in] BaseAddress        The 64 bit physical address of the memory.
+  @param[in] Length             The length of the memory allocation in bytes.
+**/
+VOID
+BuildRuntimeMemoryAllocationInfoHob (
+  VOID
+  )
+{
+  BuildGuidDataHob (
+    &gMemoryBucketInformationGuid,
+    &mRuntimeBucketHob,
+    (sizeof (PEI_MEMORY_BUCKET_INFORMATION))
+    );
+}
+
+/**
+  Function that makes pulls the memory bucket hob information locally
+  if necessary.  This is so it can be more easily referenced.
+
+  @param[in] MemBucketHob   A passed in hob that's associated with the
+                            PEI_MEMORY_BUCKET_INFORMATION GUID.
+**/
+VOID
+EFIAPI
+SyncMemoryBuckets (
+  IN VOID  *MemBucketHob
+  )
+{
+  if (!IsRuntimeMemoryInitialized () && (MemBucketHob != NULL)) {
+    SetMemoryBucketsFromHob (GET_GUID_HOB_DATA (MemBucketHob));
+  }
+}
+
+/**
+  This function returns the Memory Bucket Hob structure.
+
+  @retval Returns the Memory Bucket Hob structure.
+**/
+PEI_MEMORY_BUCKET_INFORMATION
+EFIAPI
+GetRuntimeBucketHob (
+  VOID
+  )
+{
+  return mRuntimeBucketHob;
+}
+
+/**
+  This function gets the PEI memory bucket hob and populates the locally kept data here.
+  This is to keep the memory buckets coherent.
+
+  @param[in] MemoryBuckets   The data associated with the memory bucket hob.  This gets
+                             the already allocated Memory Bucket Hob to keep coherency
+                             with the PEI memory buckets.
+**/
+VOID
+EFIAPI
+SetMemoryBucketsFromHob (
+  VOID  *MemoryBuckets
+  )
+{
+  PEI_MEMORY_BUCKET_INFORMATION  *TempStats;
+  UINTN                          Index;
+
+  if (MemoryBuckets == NULL) {
+    DEBUG ((DEBUG_ERROR, "NO PEI RUNTIME MEMORY BUCKETS YET\n"));
+    return;
+  }
+
+  TempStats = (PEI_MEMORY_BUCKET_INFORMATION *)MemoryBuckets;

```

Comment:
Can the `MemoryBuckets` parameter to the function just be of `PEI_MEMORY_BUCKET_INFORMATION *` type? If that's the only type expected, it will make the API more explicit.

Reply:
That makes sense.  Updated.